File: neetcode/blind75_rust/src/code/missing_number.rs
pub struct Solution {}

impl Solution {
    pub fn missing_number(mut nums: Vec<i32>) -> i32 {
        let eles = nums.len() as i32;
        let total = (1 + eles) * eles / 2;
        let res = total - nums.iter().fold(0, |acc, &x| acc + x);

        res
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn case_1() {
        let nums = vec![3, 0, 1];
        let expected = 2;
        let actual = Solution::missing_number(nums);

        assert_eq!(expected, actual);
    }

    #[test]
    fn case_2() {
        let nums = vec![0, 1];
        let expected = 2;
        let actual = Solution::missing_number(nums);

        assert_eq!(expected, actual);
    }

    #[test]
    fn case_3() {
        let nums = vec![9, 6, 4, 2, 3, 5, 7, 0, 1];
        let expected = 8;
        let actual = Solution::missing_number(nums);

        assert_eq!(expected, actual);
    }
}

File: python/missing_number.py
class Solution:
    def missingNumber(self, nums: list[int]) -> int:
        result: int = len(nums)
        for i in range(len(nums)):
            result ^= i ^ nums[i]
        return result


def test_missingNumber_case_1():
    # arrange
    nums: list[int] = [1, 2, 3]
    expected: int = 0

    # act
    actual: int = Solution().missingNumber(nums)

    # assert
    assert actual == expected


def test_missingNumber_case_2():
    # arrange
    nums: list[int] = [0, 2]
    expected: int = 1

    # act
    actual: int = Solution().missingNumber(nums)

    # assert
    assert actual == expected
